enhancements/operator-sdk/downstream-operator-sdk.md

```diff
@@ -0,0 +1,385 @@
+branch:
+
+* ci
+  * dockerfiles
+    * variety of Dockerfiles used to build images for running tests
+* prow.Makefile # Makefile used by CI
+* patches
+  * contain the patch files that may be carried
+* release
+  * ansible
+    * ansible_collections # contains the sync ansible collections
+    * Dockerfiles # various dockerfiles
+  * helm
+    * Dockerfiles # various dockerfiles
+  * sdk
+    * Dockerfiles # various dockerfiles
+  * scorecard-test
+    * Dockerfiles # various dockerfiles
+  * scorecard-test-kuttl
+    * Dockerfiles # various dockerfiles
+* vendor
+  * contains the dependencies for the project
+
+#### Handling Patches
+
+Typically we want to submit all bugs upstream, have them release then sync the
+release downstream. But there will be times we need to fix a problem downstream
+to resolve a bug that we can’t wait for an upstream release. The process here
+will be twofold.
+
+1. For fixes to vendored dependencies, have a patch in the patches directory
+   that will be applied to the vendor after we run go mod vendor.
+1. For fixes to mainline code, fix the mainline code directly. NOTE: these changes
+   will be lost when re-running UPSTREAM-MERGE.sh script.
+
+For either of the above, we will always submit the patches to the appropriate
+upstream repos.
+
+To generate the patches, the easiest approach is to use [`gendiff`][gendiff].
+The general steps would be any file to be changed, copy it with a meaningful
+extension and use `gendiff` to generate the patches to put in the `patch`
+directory.
+
+Let's assume `operator_installer.go` needs a fix downstream to fix a bug.
+
+1. Make backup:
+   * `cp operator_installer.go operator_installer.go.bugXXXXXX`
+1. Make changes required to fix the problem:
+   * `vim operator_installer.go`
+1. Generate patch
+   * cd ROOT_OF_PROJECT
+   * `gendiff directory_of_file .bugXXXXXX > 00_bugXXXXXX.patch`
+1. Add patch to patches directory.
+   * `cp 00_bugXXXXXX.patch patches/`
+   * `git add patches/00_bugXXXXXX.patch`
+   * `git commit -m "Bug XXXXXX: fixing ..."`
+
+Patches will be applied in order. The patches will live in the overlay branch.
+
+### Risks and Mitigation
+
+Moving forward with the automated syncing is a little high risk in the sense
+that there are some unknown issues we may encounter. We will mitigate this risk
+by keeping the existing `openshift/ocp-release-operator-sdk` repo until the
+automated `openshift/operator-sdk` repo is working. Once the cut over is made we
+can remove the `openshift/ocp-release-operator-sdk` repo.
+
+In the event that the automated repo is not ready in time for 4.7, the backup
+plan is to do a 4.7 release of OperatorSDK from the existing
+`openshift/ocp-release-operator-sdk` repo.
+
+## Design Details
+
+### Open Questions [optional]
+
+N/A
+
+### Test Plan
+
+The existing set of Operator SDK tests will be run downstream. Today we run
+tests for the Ansible and Helm operators.
```

Comment:
The upstream tests will be run downstream? Or the existing downstream tests will be run downstream?

Either way, we should document:
- what those tests are (e.g. check links, test-sanity, e2e and integration tests)
- what infrastructure is needed (e.g. kind; maybe downstream we would want to use OpenShift)
- what are the gaps that would need to be covered?

Also, if we change our tests upstream, will they be reflected downstream?

Reply:
I understand that the idea for the proposal is to keep the same upstream tests on downstream. The difference would be that we need to run the against OCP instead of Kind. Then, shows that will be required an effort to make the images work on that.

Reply:
Updated to include the Makefile targets we will run downstream. We can't run test-links because that is checking links for upstream docs. As far as tests being updated upstream; they'll get pulled in during the sync.